oxidation/libparsec/crates/core_fs/src/file_operations.rs

```diff
@@ -0,0 +1,713 @@
+        removed_ids.remove(&start_chunk.id);
+    }
+
+    // Add new buffer
+    new_chunks.push(new_chunk);
+
+    // Test stop_chunk
+    let stop_chunk = chunks
+        .get(stop_index - 1)
+        .expect("Indexes are found using binary search and hence always valid");
+    if stop_chunk.stop.get() > stop {
+        let mut new_stop_chunk = stop_chunk.clone();
+        new_stop_chunk.start = stop;
+        new_chunks.push(new_stop_chunk);
+        removed_ids.remove(&stop_chunk.id);
+    }
+
+    // Chunks after start chunk
+    new_chunks.extend(
+        chunks
+            .get(stop_index..)
+            .unwrap_or_default()
+            .iter()
+            .map(|chunk| {
+                // The same ID might appear in multiple chunks,
+                // so it's crucial that we make sure to not remove an ID
+                // that ends up being part of the new manifest
+                removed_ids.remove(&chunk.id);
+                chunk
+            })
+            .cloned(),
+    );
+
+    (new_chunks, removed_ids)
+}
+
+pub fn prepare_write(
+    manifest: &mut LocalFileManifest,
+    mut size: u64,
+    mut offset: u64,
+    timestamp: DateTime,
+) -> (Vec<WriteOperation>, HashSet<ChunkID>) {
+    let mut padding = 0;
+    let mut removed_ids = HashSet::new();
+    let mut write_operations = vec![];
+
+    // Padding
+    if offset > manifest.size {
+        padding = offset - manifest.size;
+        size += padding;
+        offset = manifest.size;
+    }
+
+    // Find proper block indexes
+    let blocksize = u64::from(manifest.blocksize);
+    let start_block = offset / blocksize;
+    let stop_block = (offset + size + blocksize - 1) / blocksize;
+
+    // Loop over blocks
+    for block in start_block..stop_block {
+        // Get substart, substop and subsize
+        let blockstart = block * blocksize;
+        let substart = max(offset, blockstart);
+        let substop = min(offset + size, blockstart + blocksize);
+        let subsize = substop
+            .checked_sub(substart)
+            .expect("Substop is always greater than substart");
+        let content_offset = substart - offset;
+
+        // Prepare new chunk
+        let new_chunk = Chunk::new(
+            substart,
+            NonZeroU64::new(substart + subsize)
+                .expect("subsize is always strictly greater than zero"),
+        );
+        write_operations.push((new_chunk.clone(), content_offset as i64 - padding as i64));
+
+        // Get the corresponding chunks
+        let new_chunks = match manifest.get_chunks(block as usize) {
+            Some(block_chunks) => {
+                let (new_chunks, more_removed_ids) =
+                    block_write(block_chunks, subsize, substart, new_chunk);
+                removed_ids.extend(more_removed_ids);
+                new_chunks
+            }
+            None => vec![new_chunk],
+        };
+
+        // Update data structures
+        if manifest.blocks.len() == block as usize {
```

Comment:
```suggestion
        if manifest.blocks.len() as u64 == block {
```
less prone to overflow the other way around